### PR TITLE
ENH: `Styler.bar` extended to allow centering about the mean, value or callable

### LINKED ISSUE #26070: Pandas style.bar(align='mid') does not make sense.
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np
N = 10
df = pd.DataFrame(np.random.normal(size=(10, N))+np.arange(N)*100, columns=range(N))
df.style.bar(align="mid", color=['pink', 'lightgreen'], axis=0)
```
#### Problem description

With align='mid' I would expect the middle to be the mean. Essentially, as a tool to visualize the variance. 

The dataframe above contains 10 columns with random numbers. To the n's column n*100 is added. So, column-wise, I would expect pretty similar output cause the mid should be the mean of each column. 

But because for columns that contain only positive or (only negative) values, the alignment is done differently. The way it is done right now you cannot see anything for e.g. column 8.

An alignment method where the mean is always used would be nice.

### PR DESCRIPTION
**Pull request by attack68**
This refactors `Styler.bar` to allow more flexible arguments to align bars, and an additional keyword `props` akin to other builtin methods that allow some customisation, demonstrated in the new user guide (image below)

- [x] closes #26070
- [x] tests added / passed
- [x] whats new

**All previous tests pass with minor formatting alterations.**

As a follow-on I should improve the original testing for this function - it is sporadic and unstructured.

Screen Shot 2021-06-29 at 14 53 56

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -2831,3 +2785,166 @@ def _highlight_between(
         else np.full(data.shape, True, dtype=bool)
     )
     return np.where(g_left & l_right, props, "")
+
+
+def _bar(
```
Comment (jreback) [inline pandas/io/formats/style.py:2790]: could consider moving style* things to a separate dir e.g. pandas/io/formats/style/* and then have flexibility to for example move some utilities to own modules. (obviously for future)

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine. anything need updating in the doc-string?

**Comment by attack68** [comment]
> looks fine. anything need updating in the doc-string?

have also done the changes to the doc string and added a `versionadded` or `versionchanged` tag for 1.4.0

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**